### ISSUE TITLE
window.orderedWindows() and window.visibleWindows() missing some windows

### ISSUE THREAD

**Issue by AustonZ** [Hammerspoon/hammerspoon#2650]
OS: Catalina 10.15.7

I've noticed that certain visible windows aren't showing up in `hs.window.orderedWindows()` and `hs.window.visibleWindows()`. [The documentation](https://www.hammerspoon.org/docs/hs.window.html#orderedWindows) states that all visible windows are supposed to be returned for both of these. Conveniently, this can be reproduced with the Hammerspoon Console window. 

Steps to reproduce:
1. Set your `init.lua` to:
    ```
    hs.console.clearConsole()
    local window = require("hs.window")
    
    print()
    print("allWindows() where window:isVisible():")
    windows = window.allWindows()
    for i = 1, #windows do
      if windows[i]:isVisible() then
        print("  " .. windows[i]:title())
      end
    end
    
    print()
    print("visibleWindows():")
    windows = window.visibleWindows()
    for i = 1, #windows do
      print("  " .. windows[i]:title())
    end
    
    print()
    print("orderedWindows():")
    windows = window.orderedWindows()
    for i = 1, #windows do
      print("  " .. windows[i]:title())
    end
    ```
2. Open the Hammerspoon Console window
3. Reload config
4. View Hammerspoon console output

Result:
"Hammerspoon Console" appears in the "hs.window.allWindows() where isVisible() is true" list. Yet the Hammerspoon Console does not appear in the `orderedWindows()` and `visibleWindows()` lists.

Expected:
"Hammerspoon Console" should appear in all 3 lists, because it is a visible window.

**Comment by dbalatero** [comment]
Came here to post this very same issue. Here's a screenshot of just Google Chrome and Hammerspoon running:

![image](https://user-images.githubusercontent.com/59429/130535303-95a9a2f2-4a2b-4c0a-99cb-0b255f28aa8a.png)

Yet when I print out `orderedWindows()`, I just get this output:

```txt
> p(hs.window.orderedWindows())
2021-08-23 20:10:54: { <userdata 1> -- hs.window: Hammerspoon/hammerspoon: Staggeringly powerful macOS desktop automation with Lua - Google Chrome (0x600003250038) }
```

**Comment by dbalatero** [comment]
cc @cmsj in case you have any quick ideas (or if this is actually expected behavior to exclude the HS console from the list!)

**Comment by chris-david-scholz** [comment]
**Root cause, plus a more serious consequence (macOS 26.6, Hammerspoon 1.1.1)**

The windows go missing because of an app-kind filter, not because of their visibility. `hs.window.visibleWindows()` only walks apps with `kind() > 0`:

```lua
-- extensions/window/window.lua
function window.visibleWindows()
  ...
    if app:kind()>0 and not app:isHidden() then ...
```

whereas `hs.window.allWindows()` uses `app:kind()>=0`. The console is missing because Hammerspoon itself has `kind() == 0`. `orderedWindows()` intersects `_orderedwinids()` with `visibleWindows()` and inherits the gap — `_orderedwinids()` itself does contain these windows.

Permalinks: [`allWindows()` filter](https://github.com/Hammerspoon/hammerspoon/blob/23e387e2805a9890066366e0ac96c71b27f0cfd5/extensions/window/window.lua#L93) · [`visibleWindows()` filter](https://github.com/Hammerspoon/hammerspoon/blob/23e387e2805a9890066366e0ac96c71b27f0cfd5/extensions/window/window.lua#L134) · [`orderedWindows()`](https://github.com/Hammerspoon/hammerspoon/blob/23e387e2805a9890066366e0ac96c71b27f0cfd5/extensions/window/window.lua#L182-L186)

This matters beyond the console: macOS's authentication prompts belong to `kind() == 0` agents as well — `com.apple.LocalAuthentication.UIAgent` (coreautha: Touch ID / password) and `com.apple.SecurityAgent` (admin rights, keychain). A focus-follows-mouse script that focuses the topmost `orderedWindows()` entry under the pointer therefore focuses the window *underneath* the Touch ID prompt, which activates another app and dismisses the prompt.

Measured while a LocalAuthentication prompt was on screen:

| check | result |
|---|---|
| agent `kind()` | 0 |
| prompt window in `hs.window._orderedwinids()` | true |
| in `hs.window.visibleWindows()` | false |
| in `hs.window.orderedWindows()` | false |

Aligning the filter with `allWindows()` (`kind() >= 0`) or documenting that `visibleWindows()` / `orderedWindows()` only cover regular (Dock) apps would both resolve the inconsistency reported here. Which would you prefer?

**Comment by dmgerman** [comment]
app:kind()>=0 is required to skip many other windows. So this is a false-positive.

the question is, what do we use instead?

**Comment by dmgerman** [comment]
> **Root cause, plus a more serious consequence (macOS 26.6, Hammerspoon 1.1.1)**
> 
> The windows go missing because of an app-kind filter, not because of their visibility. `hs.window.visibleWindows()` only walks apps with `kind() > 0`:
> 
> -- extensions/window/window.lua
> function window.visibleWindows()
>   ...
>     if app:kind()>0 and not app:isHidden() then ..

I don't think this is a bug. I think it is because you have disallowed Hammerspoon from appearing in the Dock

These are the return values of kind()

1 — normal GUI application; appears in the Dock.
0 — GUI-capable application, but not shown in the Dock. Typically menu-bar apps, agents, utilities, etc.
-1 — application is prohibited from having GUI elements (essentially a background process).

so please check if you have Hammerspoon in the dock.

**Comment by chris-david-scholz** [comment]
Thanks @dmgerman — you're right about the console. My Hammerspoon runs without a Dock icon, so it has `kind() == 0`. I checked: with `hs.dockIcon(true)` it becomes `kind() == 1` and the console shows up in `visibleWindows()` and `orderedWindows()`; with `hs.dockIcon(false)` it is missing again.

The case I actually ran into can't be fixed that way, though. The macOS authentication prompts belong to system agents (`com.apple.LocalAuthentication.UIAgent`, `com.apple.SecurityAgent`) with `kind() == 0`, and there is no Dock setting for them. While a Touch ID prompt is on screen, its agent is even the frontmost application — yet its window is missing from `orderedWindows()`.

So whether or not the filter is intended, the docs don't describe it: `visibleWindows()` is documented as returning "all windows that are visible as per `hs.window:isVisible()`", but the console window reports `isVisible() == true` without a Dock icon and is still left out. A sentence in the docs for `visibleWindows()` / `orderedWindows()` — e.g. "only includes windows of applications shown in the Dock (`kind() == 1`)" — would already help anyone who uses them to find out what is on screen.

**Comment by dmgerman** [comment]
You can try kind() > 0 

find the file under /Applications/Hammerspoon and modify it in place. Restart Hammerspoon.

if I remember correctly, the method used by HS is very slow for windows that don't have an interface (which includes those that return 0) 

So you might find that orderedWindows() and visibleWindows() slows down significantly (to more than 1 second) but it depends on what applications you run.

**Comment by chris-david-scholz** [comment]
Thanks — I assume you mean `kind() >= 0`. Rather than editing the file inside the signed app bundle, I recomputed both variants side by side from a running Hammerspoon (same primitives, nothing patched). Results on my machine (macOS 26.6, 8 apps with `kind 1`, 35 with `kind 0`; median of 5 runs):

| | today (`kind() > 0`) | `kind() >= 0` (with `SKIP_APPS`) |
|---|---|---|
| `visibleWindows()` | 5.0 ms | 22 ms |
| `orderedWindows()` | 7.3 ms | 15.3 ms |

The slowest `kind 0` app took 0.7 ms; the only outlier was 343 ms, in the very first run without `SKIP_APPS`. So here it's 2–4.5× slower, but nowhere near a second — though I agree it depends on which apps are running, which is exactly what `SKIP_APPS` is for.

And it does fix the case: with a Touch ID prompt on screen, the variant returns `com.apple.LocalAuthentication.UIAgent` first and the app behind it second, while today's `orderedWindows()` only returns the app behind it.

Since changing the default would cost everyone that time and would also change callers like `hs.grid`, how about an opt-in instead: an optional argument such as `hs.window.orderedWindows(includeAccessoryApps)` / `visibleWindows(includeAccessoryApps)`, default unchanged, `SKIP_APPS` still applied — plus the docs sentence. Would a PR along those lines be welcome?

**Comment by dmgerman** [comment]
Since Hammerspoon2 is being in active development, it might be better for you to follow it and suggest changes.

**Comment by chris-david-scholz** [comment]
Thanks for the pointer, that makes sense. For anyone landing here later: Hammerspoon 2 already behaves the way proposed above. Since cmsj/Hammerspoon2@f79d7a7d308c75bed521da16bffe7e5852ef2f1e (which closed cmsj/Hammerspoon2#201), `hs.window.allWindows()`, `visibleWindows()` and `orderedWindows()` skip only apps whose activation policy is `.prohibited` — the processes that never answer accessibility queries and caused the 1.8 s stalls in that issue — while accessory apps (`kind() == 0` in v1 terms, including the authentication agents) are included. Before that commit, `orderedWindows()` in v2 was limited to `.regular` apps, just like v1 is today.

For v1, a sentence in the docs of `visibleWindows()` / `orderedWindows()` would still help; I've worked around it in my own config. I'll follow Hammerspoon 2 from here.
